### PR TITLE
MAINT:ENH:sparse.linalg: Rewrite iterative solvers in Python

### LINKED ISSUE #5022: bicg returns last iterate, not the smallest-residue vector
To what extent have the iterative solvers actually been used on practical applications? We are trying them one by one on a fairly simple 3D acoustic equation, and can only get the _cg_-routine to work out of the box, even for very small lattices. This is most likely our faults, but the setup we use for _cg_ (leading to quite satisfactory performance) does not converge for the other solvers.

### LINKED ISSUE #6198: callback for Krylov methods
The Krylov solvers such as `scipy.sparse.linalg.cg` have the `callback` argument. I am using this to monitor convergence. According to the docs, for `scipy.sparse.linalg.cg` the input given to the `callback` is the solution vector and for `scipy.sparse.linalg.gmres` it is the residual vector. But in my version  0.17.1 `scipy.sparse.linalg.gmres` returns a float. I guess it is the 2-norm of the residual vector. Shouldn't the input given to `callback` be independent of the Krylov method?

### LINKED ISSUE #12748: Convergence issue of GMRES
My issue is about a convergence issue when using scipy.sparse.linalg.gmres routine. When using GMRES with a preconditioner, it seems that the convergence is tested by looking at the residual not the preconditioned residual. So, although the preconditioned residual is smaller than the absolute tolerance I ask, the iterations do no stop if the (not preconditioned) residual is greater than the tolerance. In order to fix this, I need to go into iterative.py and add the following lines (under the if conditional elif(ijob==3):
if presid < atol and np.linalg.norm(x) > 0:
  resid = 0.0
  break

This kind of hard coding is not something I prefer. I will appreciate if you can take a look at this issue and inform me about it.

Thanks!

#### Reproducing code example:

```
Sample code to reproduce the problem
```

#### Error message:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  ...
```

### LINKED ISSUE #15533: BUG: test failure in `test_x0_equals_Mb` with `bicgstab`
In conda-forge, we ran into new test failure for scipy 1.8, which appears only with MKL (which is available in cf only for x86), and only if the processor supports AVX512 (which azure CI only has for linux & windows), see https://github.com/conda-forge/scipy-feedstock/pull/199.

To not further delay the release of 1.8, we skipped that test for now, but it should IMO be fixed, especially as such processors are becoming more and more common (in fact, it's getting harder and harder to purposefully catch a non-AVX512 windows CI agent on azure)

The failure is in `test_x0_equals_Mb[bicgstab]` and looks as follows:

```
=================================== FAILURES ===================================
_________________________ test_x0_equals_Mb[bicgstab] __________________________
[...]/lib/python3.8/site-packages/scipy/sparse/linalg/_isolve/tests/test_iterative.py:538: in test_x0_equals_Mb
    assert_equal(info, 0)
E   AssertionError: 
E   Items are not equal:
E    ACTUAL: -11
E    DESIRED: 0
        A          = <10x10 sparse matrix of type '<class 'numpy.complex64'>'
	with 19 stored elements in Compressed Sparse Row format>
        b          = array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.])
        case       = <nonsymposdef>
        info       = -11
        solver     = <function bicgstab at 0x7fc6545a4ee0>
        sup        = <numpy.testing._private.utils.suppress_warnings object at 0x7fc62555d040>
        tol        = 1e-08
        x          = array([0.        +0.j, 0.50000358+0.j, 1.25016941+0.j, 2.12653797+0.j,
       3.0674186 +0.j, 4.03716976+0.j, 5.0158146 +0.j, 6.00288608+0.j,
       7.        +0.j, 8.        +0.j])
        x0         = 'Mb'
```
Also, at the end of the test suite, some (seemingly delayed) log output appears that is perhaps relevant:
```

   Normal return from subroutine COBYLA

   NFVALS =   50   F = 2.485185E+01    MAXCV = 1.999965E-10
   X = 4.955358E+00   6.666553E-01

 NNLS quitting on iteration count.
```

### LINKED ISSUE #17744: BUG: test_maxiter_worsening[lgmres] failed on riscv
### Describe your issue.

This test case fails on riscv64:

```
FAILED sparse/linalg/_isolve/tests/test_iterative.py::test_maxiter_worsening[lgmres] - AssertionError
```

This error looks similar to https://github.com/scipy/scipy/pull/13022 and https://github.com/scipy/scipy/commit/4383318f9a3a16e4e697a9920269a6947632b601. These errors are skipped on aarch64 and ppc64le, what about riscv64?

full build log: https://bugs.gentoo.org/attachment.cgi?id=847590

### Reproducing Code Example

```python
No code example, just test case.
```

### Error message

```shell
==================================================================================================== FAILURES =====================================================================================================_________________________________________________________________________________________ test_maxiter_worsening[lgmres] __________________________________________________________________________________________[gw1] linux -- Python 3.10.9 /var/tmp/portage/dev-python/scipy-1.10.0/work/scipy-1.10.0-python3_10/install/usr/bin/python3.10

solver = <function lgmres at 0x3fae312950>

    @pytest.mark.parametrize("solver", [
        pytest.param(gmres, marks=pytest.mark.xfail(platform.machine() == 'aarch64'
                                                    and sys.version_info[1] == 9,
                                                    reason="gh-13019")),
        qmr,
        pytest.param(lgmres, marks=pytest.mark.xfail(platform.machine() == 'ppc64le',
                                                     reason="fails on ppc64le")),
        pytest.param(cgs, marks=pytest.mark.xfail),
        pytest.param(bicg, marks=pytest.mark.xfail),
        pytest.param(bicgstab, marks=pytest.mark.xfail),
        pytest.param(gcrotmk, marks=pytest.mark.xfail),
        pytest.param(tfqmr, marks=pytest.mark.xfail)])
    def test_maxiter_worsening(solver):
        # Check error does not grow (boundlessly) with increasing maxiter.
        # This can occur due to the solvers hitting close to breakdown,
        # which they should detect and halt as necessary.
        # cf. gh-9100

        # Singular matrix, rhs numerically not in range
        A = np.array([[-0.1112795288033378, 0, 0, 0.16127952880333685],
                      [0, -0.13627952880333782+6.283185307179586j, 0, 0],
                      [0, 0, -0.13627952880333782-6.283185307179586j, 0],
                      [0.1112795288033368, 0j, 0j, -0.16127952880333785]])
        v = np.ones(4)
        best_error = np.inf
        tol = 7 if platform.machine() == 'aarch64' else 5

        for maxiter in range(1, 20):
            x, info = solver(A, v, maxiter=maxiter, tol=1e-8, atol=0)

            if info == 0:
                assert_(np.linalg.norm(A.dot(x) - v) <= 1e-8*np.linalg.norm(v))

            error = np.linalg.norm(A.dot(x) - v)
            best_error = min(best_error, error)

            # Check with slack
>           assert_(error <= tol*best_error)
E           AssertionError
A          = array([[-0.11127953+0.j        ,  0.        +0.j        ,
         0.        +0.j        ,  0.16127953+0.j        ],
       [ 0.        +0.j        , -0.13627953+6.28318531j,
         0.        +0.j        ,  0.        +0.j        ],
       [ 0.        +0.j        ,  0.        +0.j        ,
        -0.13627953-6.28318531j,  0.        +0.j        ],
       [ 0.11127953+0.j        ,  0.        +0.j        ,
         0.        +0.j        , -0.16127953+0.j        ]])
best_error = 1.414213562373095
error      = 67.94531422934396
info       = 8
maxiter    = 8
solver     = <function lgmres at 0x3fae312950>
tol        = 5
v          = array([1., 1., 1., 1.])
x          = array([-1.17813344e+15+0.j        , -6.61429022e+00-3.99819143j,
       -6.61429022e+00+3.99819143j, -8.12887631e+14+0.j        ])

../work/scipy-1.10.0-python3_10/install/usr/lib/python3.10/site-packages/scipy/sparse/linalg/_isolve/tests/test_iterative.py:487: AssertionError
```

### SciPy/NumPy/Python version information

1.10.0 1.23.5 sys.version_info(major=3, minor=10, micro=9, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by ilayn**
Contuniation of #18391 

---------------

As discussed in #18367 for `linalg.interpolative`, `iterative` solvers are also in the same state that it is almost impossible to implement any enhancements such as returning the iteration information and so on. This PR, reimplements all code in pure Python and with a bit of care to inplace operations thanks to NumPy no performance is lost.

@mdhaber @tupui Touched optimize tests for handling DperecationWarnings in these solvers and also did some PEP8 cleanup in separate commits. So should be easier to review those parts. When this goes in, I can go into nonlinear optimize code to find where these are triggered.

@h-vetinari The fixtures were generating ~1000 tests and skipping 500ish of them causing lots of `s`, so modified the fixtures such that tests are skipped and collect-time. Please have a look.

When ready, I'll remove the Fortran files and the compilation machinery as the last step.

#### Reference issue
Closes #5022 
Closes #6198
Closes #12748
Closes #15533
Closes #15693 
Closes #17744 

These ones are auxiliary signature issues, that will need addressing once this PR goes in

Related #12624 
Related #15657 
Related #15738
Related #16123 

#### Memory Lane for past discussions
#1466
#7623
#7624
#8400
#10341
#16050

#### What does this implement/fix?
- Rewrote all code in Python instead of using very old Fortran templates
- The performance is comparable and slightly faster with Python
- Removes all dependencies to [`_isolve/iterative`](https://github.com/scipy/scipy/tree/main/scipy/sparse/linalg/_isolve/iterative) Fortran code. 
- The `tol`/`atol` handling is just a mess and is waiting to be overhauled for at least 5 years. This PR will unite both issues for iterative solvers at least
- Most tests are drop-in replacement passed. Several segfaults are avoided in the meantime that was annoying the CI for a very long time sporadically.
- Not my favorite coding style but I went the Fortran files line-by-line and tried to be as faithful as possible. GMRES might have some tolerance changes according to my tests.

image

### CODE REVIEW

File: scipy/sparse/linalg/_isolve/tests/test_iterative.py [hunk 1]
```diff
@@ -57,6 +62,18 @@ def __repr__(self):
...
+class SingleTest:
+    def __init__(self, A, b, solver, casename, convergence=True):
+        self.A = A
+        self.b = b
+        self.solver = solver
+        self.name = casename + '-' + solver.__name__
+        self.convergence = convergence
+
+    def __repr__(self):
+        return f"<{self.name}>"
```
Comment (h-vetinari) [inline scipy/sparse/linalg/_isolve/tests/test_iterative.py:74]: Since we're explicitly setting `id=`, I think the repr isn't necessary anymore actually

File: scipy/sparse/linalg/_isolve/tests/test_iterative.py [hunk 2]
```diff
@@ -184,17 +201,35 @@ def __init__(self):
...
+                    tests += [SingleTest(case.A, case.b, solver, case.name)]
+        return tests
 
-cases = IterativeParams().cases
-@pytest.fixture(params=cases, ids=[x.name for x in cases], scope="session")
+
+cases = IterativeParams().generate_tests()
+
+
+@pytest.fixture(params=cases, ids=[x.name for x in cases], scope="module")
```
Comment (h-vetinari) [inline scipy/sparse/linalg/_isolve/tests/test_iterative.py:222]: Nice! 👍

File: scipy/sparse/linalg/_isolve/tests/test_iterative.py [hunk 3]
```diff
@@ -478,29 +472,35 @@ def test_maxiter_worsening(solver):
...
         # Check with slack
-        assert error <= tol * best_error
+        assert error <= slack_tol * best_error
 
 
 def test_x0_working(solver):
     # Easy problem
-    np.random.seed(1)
+    # np.random.seed(1)
+    rng = np.random.default_rng(1)
```
Comment (h-vetinari) [inline scipy/sparse/linalg/_isolve/tests/test_iterative.py:498]: Can we use a larger seed here and remove the commented out old code?

**Review by h-vetinari — COMMENTED** [review]
I only really looked at the test changes; I like them a lot! Also the commits are nicely separated[^1], good work! :)

[^1]: the diff of the first one is hard to read, but I guess that's in the nature of such a change.

PS. Thanks for moving the atol/rtol/tol-machinery to `_get_atol` 👍

File: scipy/sparse/linalg/_isolve/tests/test_iterative.py (hunk 3 above)
Comment (ilayn) [reply scipy/sparse/linalg/_isolve/tests/test_iterative.py:498]: Oh it's a brainfart. Also forgot to modify the TestGMRES random parts. I'll fix an ARPACK test first then fix these. Thanks.

**Comment by ilayn** [comment]
I bumped the milestone on this one because there are other solvers that needs attention about deprecations. We can wait another 6 months on top of 15 years I think :)

**Comment by ilayn** [comment]
I spoke to Pauli offline and went over about his preconditioning tolerance machinery. There is no problem as far as we can see. Also a tolerance initialization bug was causing some relaxation in the solution which is now fixed.

File: scipy/optimize/tests/test_nonlin.py [hunk 4]
```diff
@@ -168,6 +171,7 @@ def _check_root(self, f, method, f_tol=1e-2):
     def _check_func_fail(self, *a, **kw):
         pass
 
+    @pytest.mark.filterwarnings('ignore::DeprecationWarning')
```
Comment (j-bowhay) [inline scipy/optimize/tests/test_nonlin.py:174]: This patch removes the need to filter the deprecation warnings. It looks like there was a minor oversight in #8400 and how the tolerance is passed to the iterative solver as currently as far as I can tell the old `if 'tol' in self.method_kw:` branch is never hit. However, with this patch there is now one test that overflows :confused: 
```
diff --git a/scipy/optimize/_nonlin.py b/scipy/optimize/_nonlin.py
index c6abe1684..2812d01e0 100644
--- a/scipy/optimize/_nonlin.py
+++ b/scipy/optimize/_nonlin.py
@@ -1477,7 +1477,7 @@ class KrylovJacobian(Jacobian):
         return r
 
     def solve(self, rhs, tol=0):
-        if 'tol' in self.method_kw:
+        if 'atol' in self.method_kw:
             sol, info = self.method(self.op, rhs, **self.method_kw)
         else:
             sol, info = self.method(self.op, rhs, tol=tol, **self.method_kw)
diff --git a/scipy/optimize/tests/test_nonlin.py b/scipy/optimize/tests/test_nonlin.py
index d60665b6b..54100020a 100644
--- a/scipy/optimize/tests/test_nonlin.py
+++ b/scipy/optimize/tests/test_nonlin.py
@@ -171,7 +171,6 @@ class TestNonlin:
     def _check_func_fail(self, *a, **kw):
         pass
 
-    @pytest.mark.filterwarnings('ignore::DeprecationWarning')
     def test_problem_nonlin(self):
         for f in [F, F2, F2_lucky, F3, F4_powell, F5, F6]:
             for func in SOLVERS.values():
@@ -181,7 +180,6 @@ class TestNonlin:
                     continue
                 self._check_nonlin_func(f, func)
 
-    @pytest.mark.filterwarnings('ignore::DeprecationWarning')
     @pytest.mark.parametrize("method", ['lgmres', 'gmres', 'bicgstab', 'cgs',
                                         'minres', 'tfqmr'])
     def test_tol_norm_called(self, method):
@@ -197,7 +195,6 @@ class TestNonlin:
                              tol_norm=local_norm_func)
         assert_(self._tol_norm_used)
 
-    @pytest.mark.filterwarnings('ignore::DeprecationWarning')
     def test_problem_root(self):
         for f in [F, F2, F2_lucky, F3, F4_powell, F5, F6]:
             for meth in SOLVERS:
```

**Comment by j-bowhay** [comment]
This should probably wait for #18488 to avoid added complications there

File: scipy/optimize/tests/test_nonlin.py (hunk 4 above)
Comment (ilayn) [reply scipy/optimize/tests/test_nonlin.py:174]: Hmm, I am not sure I understand your point though. Do you mind elaborating? Is this related to ignoring the warning or something wrong with the test?

Comment (ilayn) [reply scipy/optimize/tests/test_nonlin.py:174]: Ah OK. Got it. Hopefully we can convert all solvers to accept atol, rtol before 1.12 is out hence we can unify all solvers under one common API. That `*` business is really a good idea in the signature to enforce it.

Comment (j-bowhay) [reply scipy/optimize/tests/test_nonlin.py:174]: +1 sounds like a good plan

**Comment by ilayn** [comment]
If everyone is happy, I'd like to add the Fortran code deletion commit on top and merge soon. Please let me know if you spot any issues in the meantime.

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 5]
```diff
@@ -1,496 +1,579 @@
...
+        rate of convergence, which implies that fewer iterations are needed
+        to reach a given error tolerance.
+    callback : function
+        User-supplied function to call after each iteration.  It is called
+        as callback(xk), where xk is the current solution vector.
+    tol : float, optional, deprecated
+
+        .. deprecated 1.11.0
+           `cg` keyword argument `tol` is deprecated in favor of `rtol` and
+           will be removed in SciPy 1.13.0.
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_isolve/iterative.py:339]:
```suggestion
        .. deprecated 1.12.0
           `cg` keyword argument `tol` is deprecated in favor of `rtol` and
           will be removed in SciPy 1.14.0.
```

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 6]
```diff
@@ -552,13 +618,29 @@ def gmres(A, b, x0=None, tol=1e-5, restart=None, maxiter=None, M=None, callback=
...
 
         .. deprecated:: 0.11.0
-           `gmres` keyword argument `restrt` is deprecated infavour of
+           `gmres` keyword argument `restrt` is deprecated in favor of
            `restart` and will be removed in SciPy 1.12.0.
+    tol : float, optional, deprecated
+
+        .. deprecated 1.11.0
+           `gmres` keyword argument `tol` is deprecated in favor of `rtol` and
+           will be removed in SciPy 1.13.0
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_isolve/iterative.py:634]:
```suggestion
        .. deprecated 1.12.0
           `gmres` keyword argument `tol` is deprecated in favor of `rtol` and
           will be removed in SciPy 1.14.0
```

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 7]
```diff
@@ -782,6 +888,21 @@ def qmr(A, b, x0=None, tol=1e-5, maxiter=None, M1=None, M2=None, callback=None,
     callback : function
         User-supplied function to call after each iteration.  It is called
         as callback(xk), where xk is the current solution vector.
+    tol : float, optional, deprecated
+
+        .. deprecated 1.11.0
+           `qmr` keyword argument `tol` is deprecated in favor of `rtol` and
+           will be removed in SciPy 1.13.0.
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_isolve/iterative.py:895]:
```suggestion
        .. deprecated 1.12.0
           `qmr` keyword argument `tol` is deprecated in favor of `rtol` and
           will be removed in SciPy 1.14.0.
```

**Review by j-bowhay — CHANGES_REQUESTED** [review]
Minor tweak needed due to bumping the milestone

**Review by j-bowhay — CHANGES_REQUESTED** [review]
Minor tweak needed due to bumping the milestone

File: scipy/optimize/tests/test_nonlin.py (hunk 4 above)
Comment (h-vetinari) [reply scipy/optimize/tests/test_nonlin.py:174]: Sidenote: In my notifications, I got a long comment from Jake here that is now gone. I'm not sure if it was deleted (cannot find it in the edits), but that makes things hard to follow. It's unavoidable to edit comments occasionally, but can we avoid deleting them?

Comment (j-bowhay) [reply scipy/optimize/tests/test_nonlin.py:174]: Sorry, I wrote that comment and then realised I had miss understood part of the code in `scipy/optimize/_nonlin.py` and that the comment was going to cause unnecessary confusion so deleted it.

**Review by h-vetinari — COMMENTED** [review]
Cannot re-review in depth, but the test changes look benign, which means that the implementation should be doing what it's supposed to. I trust you on this. 👍

**Review by rgommers — COMMENTED** [review]
This looks great @ilayn! I had a look now, and can review in more detail and get this merged - just can't promise it happens this week, because I have a trip starting tomorrow.

One issue I just noticed with the `tol` deprecation (which came up in an unrelated PR earlier today): it won't trigger if a user writes something like: `bicg(A, b, x0, None, 1000, M)`. To address that, the signature can be changed to use `tol=_NoValue` with `_NoValue` being a constant defined in the file. And then also emit the deprecation warning if the user passes `None`.

> The performance is comparable and slightly faster with Python

I didn't see any benchmark results here or on the initial PR - may be worth updating the PR description with whatever results you have?

**Comment by rgommers** [comment]
> It's a bit better than that. if somebody was not using tol before, they don't get any warning because now `tol=None` and `rtol` is the new `tol`. Hence no disruption.

That is a good idea in principle, however my point was that you can then not remove the keyword. Use of a positional `None` should be noisy first before it can be ripped out.

**Comment by ilayn** [comment]
Here are some very funky results :) 

image


I have no idea what is going on with cgs or gmres but in general we are in good shape The difference looks big but the scale is zoomed hence it's not that big. 

The expensive parts of these solvers are the matvecs and preconditioner ops but they were already being done on the python side anyways hence the performance is pretty much the same if we don't stray away from inplace NumPy ops. 

I'll check on GMRES one more time to see if I forgot some inplace linalg function but I'd take this PR against that sphagetti even if we lost 20% to be honest.

**Comment by ilayn** [comment]
> That is a good idea in principle, however my point was that you can then not remove the keyword. Use of a positional None should be noisy first before it can be ripped out.

Ah you mean you don't want to yank out a keyword from the middle to disturb the positional arguments if I understand correctly. But I think that kind of breakage is justified in this situation. Anyways let's continue when you are back 😃 I'll also need to reset myself from all this FORTRAN radiation I've been exposed lately. .

**Comment by ilayn** [comment]
Just as a reminder, how should we proceed with the tol keyword here? Don't know a separate PR might be a better strategy to deal with it but I guess @rgommers saying is a better way to go once and for all.

**Comment by h-vetinari** [comment]
> Just as a reminder, how should we proceed with the tol keyword here?

The deprecation warning from #8400 (in `gcrotmk` & `lgmres`) has been warning for 5 years not to use `tol=` and recommends to use `atol=`.

In fact, it's warning as soon as `atol=None`, and since that argument's at the very end of the [signature](https://github.com/scipy/scipy/blob/v1.10.1/scipy/sparse/linalg/_isolve/_gcrotmk.py#L182-L184), someone would have to fill _all_ arguments positionally _and_ set `atol` to something non-`None` to both: not see that warning and be broken by removal of `tol`.

I think this is very unlikely usage, and considering the length of time this has been warning, it'd be defensible IMO to just remove `tol`.

Haven't looked how many other functions are affected by this though, and I can also see the argument that it'd be cleaner to deprecate positional usage completely (and then switch to keyword-only arguments).

**Comment by ilayn** [comment]
@ @h-vetinari @rgommers @j-bowhay Should we merge this and deal with the `_NoValue()` right after this? I am a bit worried that it's too crowded in this PR to find all the occurences of the `tol=None` usage. I can submit something very quickly for that separately.

**Comment by h-vetinari** [comment]
Fine by me[^1]

[^1]: not that I wouldn't prefer it to be available in a separate commit or PR before merging, but I don't want to impede your progress on this, and I trust you'll follow up.

**Comment by ilayn** [comment]
OK, I'll open the PR in a bit.
